Move doc-values classes needed by ST_INTERSECTS to server

craigtaverner opened the pull request:
These classes are needed by ESQL spatial queries, and are not licensed in a way that prevents this move. Since they depend on lucene it is not possible to move them to a library. Instead they are moved to be co-located with the GeoPoint doc-values classes that already exist in server.

craigtaverner committed:
Move doc-values classes needed by ST_INTERSECTS to server

This classes are needed by ESQL spatial queries, and are not licensed in a way that prevents this move.
Since they depend on lucene it is not possible to move them to a library.
Instead they are moved to be co-located with the GeoPoint doc-values classes that already exist in server.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

iverase reviewed (changes requested):
Can we please move the under a package in ~~`org.apache.lucene`~~  `org.elasticsearch.lucene`? They are really implementing some lucene functionality.

craigtaverner committed:
Moved to lucene package org.elasticsearch.lucene.spatial

craigtaverner committed:
Moved Geo/ShapeDocValuesQuery to server because it is Lucene specific

And this gives us access to these classes from ESQL for lucene-pushdown of spatial queries.

iverase commented on x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/fielddata/Tile2DVisitorTests.java:
```diff
@@ -287,4 +290,40 @@ static void assertRelation(GeometryDocValueReader reader, Extent extent, GeoRela
         assertThat(tile2DVisitor.relation(), in(expectedRelation));
     }
+
+    private static class TestCoordinateEncoder implements CoordinateEncoder {
```
nice!

iverase reviewed (approved):
lgtm

craigtaverner commented:
@elasticsearchmachine run elasticsearch-ci/part-1
